### PR TITLE
feat: add json patches

### PR DESCRIPTION
**Pull request by dnguy078**
- Add ability to apply patches to PodSpecs via [JSON Patches ](https://datatracker.ietf.org/doc/html/rfc6902/) for PodAccessTemplates

### CODE REVIEW

File: internal/api/v1alpha1/pod_spec_mutation_config.go [hunk 1]
```diff
@@ -73,6 +75,10 @@ type PodTemplateSpecMutationConfig struct {
 	// +kubebuilder:default:=false
 	PurgeAnnotations bool `json:"purgeAnnotations,omitempty"`
 
+	// PatchSpecOperations contains a list of JSON patch operations to apply to the PodSpec.
+	// [`JSONPatch`](https://www.rfc-editor.org/rfc/rfc6902.html)
+	PatchSpecOperations []map[string]string `json:"patchSpecOperations,omitempty"`
```
Comment (diranged) [inline internal/api/v1alpha1/pod_spec_mutation_config.go:80]: Hmm. Is there not a better type we can use here? I could put in an easily mis-typed or incorrect patch:

```
    patchSpecOperations:
      - notAnOp: replace
        mutatePath: '/spec/containers/0/name'
        badValue: dennis
```

File: internal/api/v1alpha1/pod_spec_mutation_config.go [hunk 2]
```diff
@@ -282,5 +288,35 @@ func (c *PodTemplateSpecMutationConfig) PatchPodTemplateSpec(
...
+		if err != nil {
+			return n, err
+		}
+
+		var podTemplateSpec corev1.PodTemplateSpec
+		if err := json.Unmarshal(modified, &podTemplateSpec); err != nil {
+			return n, err
+		}
+
+		logger.V(1).Info(fmt.Sprintf("Applied JSON patch operations, containerID: %d, patches: %+v", defContainerID, string(p)))
```
Comment (diranged) [inline internal/api/v1alpha1/pod_spec_mutation_config.go:317]: Does `defContainerID` even apply here? The patches apply to the entire `corev1.PodTemplateSpec` - which means the patch can apply to any container in the spec.

Comment (diranged) [reply internal/api/v1alpha1/pod_spec_mutation_config.go:317]: I don't know if we should print out the entire patch ... at least not in string format. Can you show me what this looks like if you were to put in several patches?

**Review by diranged — COMMENTED** [review]
@dnguy078 This is great... I have a few issues I'd like to resolve in there, but overall its clean, simple and very flexible.

File: internal/api/v1alpha1/pod_spec_mutation_config.go (hunk 1 above)
Comment (dnguy078) [reply internal/api/v1alpha1/pod_spec_mutation_config.go:80]: I'll see if we can do sort of validation or a better type here. The current third party library used in this PR uses just `[]bytes`.

Comment (diranged) [reply internal/api/v1alpha1/pod_spec_mutation_config.go:80]: Perhaps we can build our own JSNOPatch type - and also use https://github.com/evanphx/json-patch/blob/master/v5/patch.go#L905-L918

Comment (dnguy078) [reply internal/api/v1alpha1/pod_spec_mutation_config.go:80]: Yea i tried to create own JSON Patch type but looks using `interface{}` for `value` causes issues from `controller-tools` [ref](https://github.com/kubernetes-sigs/controller-tools/pull/126). We could update to `string` if anything? 

```
type JSONPatchOperation struct {
	// +kubebuilder:validation:Required
	Operation string `json:"op"`
	// +kubebuilder:validation:Required
	Path string `json:"path"`
	// +kubebuilder:validation:Optional
	Value interface{} `json:"value,omitempty"`
}
```

Comment (dnguy078) [reply internal/api/v1alpha1/pod_spec_mutation_config.go]: Ultimately went with 
```
// JSONPatchOperation represents a JSON Patch operation defined in https://www.rfc-editor.org/rfc/rfc6902.html
type JSONPatchOperation struct {
	// +kubebuilder:validation:Required
	Operation string `json:"op"`
	// +kubebuilder:validation:Required
	Path string `json:"path"`
	// +kubebuilder:validation:Optional
	Value intstr.IntOrString `json:"value,omitempty"`
}
```

manually verified basic validations as well...

``` 
➜  examples git:(add-json-patches-feat) ✗ k apply -f pod_access_template.yaml

The PodAccessTemplate "deployment-example" is invalid: spec.controllerTargetMutationConfig.patchSpecOperations[0].op: Invalid value: "integer": spec.controllerTargetMutationConfig.patchSpecOperations[0].op in body must be of type string: "integer"
```

validatePatch gets called explicitly by `DecodePatch` which we already call as well.

**Comment by diranged** [comment]
@dnguy078 What do you think about https://github.com/diranged/oz/pull/234/files merging into this?

**Comment by dnguy078** [comment]
> @dnguy078 What do you think about https://github.com/diranged/oz/pull/234/files merging into this?

Looks good. I’ll need to bump my local go version to 1.21 to fix the go.mod

File: internal/api/v1alpha1/pod_spec_mutation_config.go (hunk 2 above)
Comment (dnguy078) [reply internal/api/v1alpha1/pod_spec_mutation_config.go]: I ended up removing this logging. I don't believe we will need it.

**Pull request merged**